contracts/src/v0.8/dev/automation/upkeeps/ProxyBalanceMonitor.sol

```diff
@@ -0,0 +1,287 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.4;
+
+import "../../../ConfirmedOwner.sol";
+import "../../../interfaces/automation/KeeperCompatibleInterface.sol";
+import "../../../vendor/openzeppelin-solidity/v4.7.0/contracts/security/Pausable.sol";
+import "../../../vendor/openzeppelin-solidity/v4.7.0/contracts/token/ERC20/IERC20.sol";
+import "../../../vendor/openzeppelin-solidity/v4.7.3/contracts/utils/structs/EnumerableSet.sol";
+
+interface IAggregatorProxy {
+  function aggregator() external view returns (address);
+}
+
+interface IAggregator {
+  function linkAvailableForPayment() external view returns (int256 availableBalance);
+
+  function transmitters() external view returns (address[] memory);
+}
+
+/**
+ * @title The ProxyBalanceMonitor contract.
+ * @notice A keeper-compatible contract that monitors proxy-gated aggregators and funds them with LINK
+ * @dev with 30 addresses as the MAX_PERFORM, the measured max gas usage of performUpkeep is around 2M
+ * therefore, we recommend an upkeep gas limit of 3M (this has a 33% margin of safety). Although, nothing
+ * prevents us from using 5M gas and increasing MAX_PERFORM, 30 seems like a reasonable batch size that
+ * is probably plenty for most needs.
+ * @dev with 140 addresses as the MAX_CHECK, the measured max gas usage of checkUpkeep is around 3.5M,
+ * which is 30% below the 5M limit.
+ * Note that testing conditions DO NOT match live chain gas usage, hence the margins. Change
+ * at your own risk!!!
+ * @dev some areas for improvement / acknowledgement of limitations:
+ * * validate that all addresses conform to interface when adding them to the watchlist
+ * * this is a "trusless" upkeep, meaning it does not trust the caller of performUpkeep;
+     we could save a fair amount of gas and re-write this upkeep for use with Automation v2.0+,
+     which has significantly different trust assumptions
+ */
+contract ProxyBalanceMonitor is ConfirmedOwner, Pausable, KeeperCompatibleInterface {
+  using EnumerableSet for EnumerableSet.AddressSet;
+
+  event FundsWithdrawn(uint256 amountWithdrawn, address payee);
+  event TopUpSucceeded(address indexed topUpAddress);
+  event TopUpBlocked(address indexed topUpAddress);
+  event WatchlistUpdated();
+
+  error InvalidWatchList();
+  error DuplicateAddress(address duplicate);
+
+  uint256 public constant MAX_PERFORM = 30; // max number to addresses to top up in a single batch
+  uint256 public constant MAX_CHECK = 140; // max number of upkeeps to check (need to fit in 5M gas limit)
+  IERC20 public immutable LINK_TOKEN;
+
+  EnumerableSet.AddressSet private s_watchList;
+  uint256 private s_minBalance;
+  uint256 private s_topUpAmount;
+
+  /**
+   * @param linkTokenAddress the LINK token address
+   * @param minBalance the minimum balance an aggregator can have before initiating a top up
```

Comment:
Aggregators can potentially have a negative balance - should this be an `int256` so that there is no comparing an unsigned integer to a signed integer?

Reply:
i think this is handled in line 272 `if (balance < 0 || uint256(balance) > s_minBalance) {`